# **Exploración y preprocesamiento de un conjunto de datos con Pandas**

El objetivo de este ejercicio es repasar la biblioteca **Pandas**, para preprocesar y analizar los conjuntos de datos antes de aplicarles técnicas de visualización de datos.

## **El conjunto de datos: Inmigración a Canadá entre 1980 y 2013**

Fuente del conjunto de datos: [Flujos migratorios internacionales hacia y desde determinados países — Revisión de 2015](https://www.un.org/development/desa/pd/data/international-migration-flows).

El conjunto de datos contiene información anual sobre los flujos de inmigrantes internacionales tal y como los registran los países de destino. Los datos recogen tanto las entradas como las salidas según el lugar de nacimiento, la nacionalidad o el lugar de residencia anterior o posterior, tanto para extranjeros como para nacionales. Esta revisión presenta datos correspondientes a 45 países; nos centraremos en los datos de inmigración de Canadá.

![Vista previa de los datos](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/Module%201/images/DataSnapshot.png)

 El conjunto de datos sobre inmigración en Canadá se puede descargar desde <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx">aquí</a>.

## **Conceptos básicos de *pandas***

## 1. Instalamos **openpyxl**, módulo que *pandas* necesita para leer archivos de Excel.


In [2]:
!pip install --upgrade openpyxl

Importamos dos módulos clave para el análisis de datos: *pandas* y *numpy*.

In [3]:
import numpy as np  # útil para muchos cálculos científicos en Python
import pandas as pd # biblioteca principal de estructuras de datos

Descargamos e importamos el conjunto de datos principal sobre inmigración canadiense, utilizamos el método read_excel() de pandas.

In [4]:
df_can = pd.read_excel(
    'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx',
    sheet_name='Canada by Citizenship',
    skiprows=range(20),
    skipfooter=2)

print('Datos importados a un dataframe de pandas')

Datos importados a un dataframe de pandas


Vemos las cinco primeras filas del conjunto de datos utilizando la función `head()` y las cinco últimas usando `tail()`.

In [5]:
df_can.head()

,Type,Coverage,OdName,AREA,AreaName,REG,RegName,DEV,DevName,1980,...,2004,2005,2006,2007,2008,2009,2010,2011,2012,2013
0,Immigrants,Foreigners,Afghanistan,935,Asia,5501,Southern Asia,902,Developing regions,16,...,2978,3436,3009,2652,2111,1746,1758,2203,2635,2004
1,Immigrants,Foreigners,Albania,908,Europe,925,Southern Europe,901,Developed regions,1,...,1450,1223,856,702,560,716,561,539,620,603
2,Immigrants,Foreigners,Algeria,903,Africa,912,Northern Africa,902,Developing regions,80,...,3616,3626,4807,3623,4005,5393,4752,4325,3774,4331
3,Immigrants,Foreigners,American Samoa,909,Oceania,957,Polynesia,902,Developing regions,0,...,0,0,1,0,0,0,0,0,0,0
4,Immigrants,Foreigners,Andorra,908,Europe,925,Southern Europe,901,Developed regions,0,...,0,0,1,1,0,0,0,0,1,1


In [6]:
df_can.tail()

,Type,Coverage,OdName,AREA,AreaName,REG,RegName,DEV,DevName,1980,...,2004,2005,2006,2007,2008,2009,2010,2011,2012,2013
190,Immigrants,Foreigners,Viet Nam,935,Asia,920,South-Eastern Asia,902,Developing regions,1191,...,1816,1852,3153,2574,1784,2171,1942,1723,1731,2112
191,Immigrants,Foreigners,Western Sahara,903,Africa,912,Northern Africa,902,Developing regions,0,...,0,0,1,0,0,0,0,0,0,0
192,Immigrants,Foreigners,Yemen,935,Asia,922,Western Asia,902,Developing regions,1,...,124,161,140,122,133,128,211,160,174,217
193,Immigrants,Foreigners,Zambia,903,Africa,910,Eastern Africa,902,Developing regions,11,...,56,91,77,71,64,60,102,69,46,59
194,Immigrants,Foreigners,Zimbabwe,903,Africa,910,Eastern Africa,902,Developing regions,72,...,1450,615,454,663,611,508,494,434,437,407


Utilizamos el método `info()` para obtener un breve resumen del marco de datos.

In [7]:
df_can.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Columns: 43 entries, Type to 2013
dtypes: int64(37), object(6)
memory usage: 65.6+ KB


In [8]:
df_can.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 43 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Type      195 non-null    object
 1   Coverage  195 non-null    object
 2   OdName    195 non-null    object
 3   AREA      195 non-null    int64 
 4   AreaName  195 non-null    object
 5   REG       195 non-null    int64 
 6   RegName   195 non-null    object
 7   DEV       195 non-null    int64 
 8   DevName   195 non-null    object
 9   1980      195 non-null    int64 
 10  1981      195 non-null    int64 
 11  1982      195 non-null    int64 
 12  1983      195 non-null    int64 
 13  1984      195 non-null    int64 
 14  1985      195 non-null    int64 
 15  1986      195 non-null    int64 
 16  1987      195 non-null    int64 
 17  1988      195 non-null    int64 
 18  1989      195 non-null    int64 
 19  1990      195 non-null    int64 
 20  1991      195 non-null    int64 
 21  1992      195 no

In [9]:
df_can.columns

Index([    'Type', 'Coverage',   'OdName',     'AREA', 'AreaName',      'REG',
        'RegName',      'DEV',  'DevName',       1980,       1981,       1982,
             1983,       1984,       1985,       1986,       1987,       1988,
             1989,       1990,       1991,       1992,       1993,       1994,
             1995,       1996,       1997,       1998,       1999,       2000,
             2001,       2002,       2003,       2004,       2005,       2006,
             2007,       2008,       2009,       2010,       2011,       2012,
             2013],
      dtype='object')

In [10]:
df_can.index

RangeIndex(start=0, stop=195, step=1)

In [11]:
print(type(df_can.columns))
print(type(df_can.index))

<class 'pandas.core.indexes.base.Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


In [12]:
df_can.columns.tolist()

['Type',
 'Coverage',
 'OdName',
 'AREA',
 'AreaName',
 'REG',
 'RegName',
 'DEV',
 'DevName',
 1980,
 1981,
 1982,
 1983,
 1984,
 1985,
 1986,
 1987,
 1988,
 1989,
 1990,
 1991,
 1992,
 1993,
 1994,
 1995,
 1996,
 1997,
 1998,
 1999,
 2000,
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013]

In [13]:
print(type(df_can.columns.tolist()))
print(type(df_can.index.tolist()))

<class 'list'>
<class 'list'>


In [14]:
# tamaño del dataframe (filas, columnas)
df_can.shape

(195, 43)

**Nota**: Los principales tipos almacenados en los objetos de *pandas* son `float`, `int`, `bool`, `datetime64[ns]`, `datetime64[ns, tz]`, `timedelta[ns]`, `category` y `object` (cadena de caracteres). Además, estos tipos de datos tienen distintos tamaños de elemento, por ejemplo, `int64` e `int32`.

# **LIMPIEZA**

Limpiemos el conjunto de datos para eliminar columnas innecesarias. Utilizamos `drop()`:

In [15]:
# En pandas axis=0 representa las filas (por defecto) y axis=1 representa las columnas.
df_can.drop(['AREA', 'REG', 'DEV', 'Type', 'Coverage'], axis=1, inplace=True)
df_can.head(2)

,OdName,AreaName,RegName,DevName,1980,1981,1982,1983,1984,1985,...,2004,2005,2006,2007,2008,2009,2010,2011,2012,2013
0,Afghanistan,Asia,Southern Asia,Developing regions,16,39,39,47,71,340,...,2978,3436,3009,2652,2111,1746,1758,2203,2635,2004
1,Albania,Europe,Southern Europe,Developed regions,1,0,0,0,0,0,...,1450,1223,856,702,560,716,561,539,620,603


Cambiamos el nombre de las columnas para que tengan sentido. Podemos utilizar  `rename()` pasando un diccionario con los nombres antiguos y los nuevos:

In [16]:
df_can.rename(columns={'OdName':'Country', 'AreaName':'Continent', 'RegName':'Region'}, inplace=True)
df_can.columns

Index([  'Country', 'Continent',    'Region',   'DevName',        1980,
              1981,        1982,        1983,        1984,        1985,
              1986,        1987,        1988,        1989,        1990,
              1991,        1992,        1993,        1994,        1995,
              1996,        1997,        1998,        1999,        2000,
              2001,        2002,        2003,        2004,        2005,
              2006,        2007,        2008,        2009,        2010,
              2011,        2012,        2013],
      dtype='object')

Obtenemos la cantidad total de inmigrantes por país, usamos dos métodos.

In [17]:
df_can['Total'] = df_can.sum(axis=1, numeric_only=True)
df_can['Total']

,Total
0,58639
1,15699
2,69439
3,6
4,15
...,...
190,97146
191,2
192,2985
193,1677


In [18]:
# 1. Creamos la lista de años como números enteros (de 1980 a 2013)
years = list(range(1980, 2014))

# 2. Sumamos solo esas columnas numéricas
df_can['Total1'] = df_can[years].sum(axis=1)

# 3. Mostramos el resultado
df_can['Total1']

,Total1
0,58639
1,15699
2,69439
3,6
4,15
...,...
190,97146
191,2
192,2985
193,1677


Podemos comprobar cuántos objetos nulos hay en el conjunto de datos de la siguiente manera:

In [19]:
df_can.isnull().sum()

,0
Country,0
Continent,0
Region,0
DevName,0
1980,0
1981,0
1982,0
1983,0
1984,0
1985,0


Por último, veamos un breve resumen de cada columna de nuestro dataframe utilizando el método describe().

In [20]:
df_can.describe()

,1980,1981,1982,1983,1984,1985,1986,1987,1988,1989,...,2006,2007,2008,2009,2010,2011,2012,2013,Total,Total1
count,195.000000,195.000000,195.000000,195.000000,195.000000,195.000000,195.000000,195.000000,195.000000,195.000000,...,195.000000,195.000000,195.000000,195.000000,195.000000,195.000000,195.000000,195.000000,195.000000,195.000000
mean,508.394872,566.989744,534.723077,387.435897,376.497436,358.861538,441.271795,691.133333,714.389744,843.241026,...,1266.958974,1191.820513,1246.394872,1275.733333,1420.287179,1262.533333,1313.958974,1320.702564,32867.451282,32867.451282
std,1949.588546,2152.643752,1866.997511,1204.333597,1198.246371,1079.309600,1225.576630,2109.205607,2443.606788,2555.048874,...,3926.717747,3443.542409,3694.573544,3829.630424,4462.946328,4030.084313,4247.555161,4237.951988,91785.498686,91785.498686
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,1.000000
25%,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.500000,0.500000,1.000000,1.000000,...,25.000000,31.000000,31.000000,36.000000,40.500000,37.500000,42.500000,45.000000,952.000000,952.000000
50%,13.000000,10.000000,11.000000,12.000000,13.000000,17.000000,18.000000,26.000000,34.000000,44.000000,...,218.000000,198.000000,205.000000,214.000000,211.000000,179.000000,233.000000,213.000000,5018.000000,5018.000000
75%,251.500000,295.500000,275.000000,173.000000,181.000000,197.000000,254.000000,434.000000,409.000000,508.500000,...,842.000000,899.000000,934.500000,888.000000,932.000000,772.000000,783.000000,796.000000,22239.500000,22239.500000
max,22045.000000,24796.000000,20620.000000,10015.000000,10170.000000,9564.000000,9470.000000,21337.000000,27359.000000,23795.000000,...,33848.000000,28742.000000,30037.000000,29622.000000,38617.000000,36765.000000,34315.000000,34129.000000,691904.000000,691904.000000


## **Nivel intermedio: Indexación y selección (slicing)**

### **Seleccionar columna**
**Hay dos formas de filtrar por el nombre de una columna:**

Método 1: Rápido y sencillo, pero solo funciona si el nombre de la columna NO contiene espacios ni caracteres especiales.
```python
    df.column_name               # devuelve una serie
```

Método 2: Más robusto y permite filtrar por varias columnas.

```python
    df[“column”]                  # devuelve una serie
```

```python
    df[[“column 1”, “column 2”]]  # devuelve un dataframe
```
---

In [21]:
df_can['Country']

,Country
0,Afghanistan
1,Albania
2,Algeria
3,American Samoa
4,Andorra
...,...
190,Viet Nam
191,Western Sahara
192,Yemen
193,Zambia


Intentemos filtrar la lista de países («País») y los datos correspondientes a los años 1980-1985.

In [22]:
df_can[['Country', 1980, 1981, 1982, 1983, 1984, 1985]] # devuelve un dataframe
# fíjate en que «Country» es una cadena de caracteres y los años son números enteros.
# Para mantener la coherencia, más adelante convertiremos todos los nombres de columna a cadenas de caracteres.

,Country,1980,1981,1982,1983,1984,1985
0,Afghanistan,16,39,39,47,71,340
1,Albania,1,0,0,0,0,0
2,Algeria,80,67,71,69,63,44
3,American Samoa,0,1,0,0,0,0
4,Andorra,0,0,0,0,0,0
...,...,...,...,...,...,...,...
190,Viet Nam,1191,1829,2162,3404,7583,5907
191,Western Sahara,0,0,0,0,0,0
192,Yemen,1,2,1,6,0,18
193,Zambia,11,17,11,7,16,9


### **Seleccionar una fila**

Hay dos formas principales de seleccionar filas:

```python
    df.loc[label]    # filtra por las etiquetas del índice o la columna
    df.iloc[index]   # filtra por las posiciones del índice o la columna
```

El índice predeterminado del conjunto de datos es un rango numérico que va de 0 a 194, esto dificulta realizar una consulta por país concreto. Por ejemplo, para buscar datos sobre Japón, necesitamos conocer el valor de índice correspondiente.

Esto se soluciona estableciendo la columna «Country» como índice mediante el método `set_index()`.

In [23]:
df_can.set_index('Country', inplace=True)
# Consejo: Lo contrario de «set» es «reset». Por lo tanto, para restablecer el índice, podemos utilizar df_can.reset_index()

In [24]:
df_can.head(3)

,Continent,Region,DevName,1980,1981,1982,1983,1984,1985,1986,...,2006,2007,2008,2009,2010,2011,2012,2013,Total,Total1
Country,,,,,,,,,,,,,,,,,,,,,
Afghanistan,Asia,Southern Asia,Developing regions,16,39,39,47,71,340,496,...,3009,2652,2111,1746,1758,2203,2635,2004,58639,58639
Albania,Europe,Southern Europe,Developed regions,1,0,0,0,0,0,1,...,856,702,560,716,561,539,620,603,15699,15699
Algeria,Africa,Northern Africa,Developing regions,80,67,71,69,63,44,69,...,4807,3623,4005,5393,4752,4325,3774,4331,69439,69439


Veamos el número de inmigrantes procedentes de Japón (fila 87) para los siguientes casos:
1. Los datos completos de la fila (todas las columnas);
2. Para el año 2013;
3. Para los años 1980 a 1985.

In [25]:
# 1. the full row data (all columns)
df_can.loc['Japan']

,Japan
Continent,Asia
Region,Eastern Asia
DevName,Developed regions
1980,701
1981,756
1982,598
1983,309
1984,246
1985,198
1986,248


In [26]:
# método alternativo
df_can.iloc[87]

,Japan
Continent,Asia
Region,Eastern Asia
DevName,Developed regions
1980,701
1981,756
1982,598
1983,309
1984,246
1985,198
1986,248


In [27]:
df_can[df_can.index == 'Japan']

,Continent,Region,DevName,1980,1981,1982,1983,1984,1985,1986,...,2006,2007,2008,2009,2010,2011,2012,2013,Total,Total1
Country,,,,,,,,,,,,,,,,,,,,,
Japan,Asia,Eastern Asia,Developed regions,701,756,598,309,246,198,248,...,1212,1250,1284,1194,1168,1265,1214,982,27707,27707


In [28]:
# 2. for year 2013
df_can.loc['Japan', 2013]

np.int64(982)

In [29]:
# método alternativo
# el año 2013 es la última columna, con un índice posicional de 36
df_can.iloc[87, 36]

np.int64(982)

In [30]:
# 3. for years 1980 to 1985
df_can.loc['Japan', [1980, 1981, 1982, 1983, 1984, 1984]]

,Japan
1980,701
1981,756
1982,598
1983,309
1984,246
1984,246


In [31]:
# Alternative Method
df_can.iloc[87, [3, 4, 5, 6, 7, 8]]

,Japan
1980,701
1981,756
1982,598
1983,309
1984,246
1985,198


Veamos el número de inmigrantes procedentes de **Haití** en los siguientes casos: <br>1. Los datos completos de la fila (todas las columnas) <br>2. Para el año 2000 <br>3. Para los años 1990 a 1995

In [32]:
# 1. Los datos completos de la fila (todas las columnas)
df_can.loc['Haiti']

,Haiti
Continent,Latin America and the Caribbean
Region,Caribbean
DevName,Developing regions
1980,1666
1981,3692
1982,3498
1983,2860
1984,1418
1985,1321
1986,1753


In [33]:
# o bien
df_can[df_can.index == 'Haiti']

,Continent,Region,DevName,1980,1981,1982,1983,1984,1985,1986,...,2006,2007,2008,2009,2010,2011,2012,2013,Total,Total1
Country,,,,,,,,,,,,,,,,,,,,,
Haiti,Latin America and the Caribbean,Caribbean,Developing regions,1666,3692,3498,2860,1418,1321,1753,...,1619,1598,2491,2080,4744,6503,5868,4152,84813,84813


In [34]:
# 2. Para el año 2000
df_can.loc['Haiti', 2000]

np.int64(1631)

In [35]:
#  3. Para los años 1990 a 1995
df_can.loc['Haiti', [1990, 1991, 1992, 1993, 1994, 1995]]

,Haiti
1990,2379
1991,2829
1992,2399
1993,3655
1994,2100
1995,2014


Los nombres de columna que son números enteros (como los años) pueden dar lugar a cierta confusión. Por ejemplo, cuando nos referimos al año 2013, se podría confundir con el índice posicional número 2013.

Para evitar esta ambigüedad, convirtamos los nombres de columna en cadenas de caracteres: de «1980» a «2013».

In [36]:
df_can.columns = list(map(str, df_can.columns))
[print (type(x)) for x in df_can.columns.values]

<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>


[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None]

Dado que hemos convertido los años en cadenas de caracteres, declaremos una variable que nos permita acceder fácilmente a todo el rango de años:

In [37]:
years = list(map(str, range(1980, 2014)))
years

['1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013']

**Ejercicio:** Crea una lista llamada «year» utilizando la función `map` para los años comprendidos entre 1990 y 2013. <br>A continuación, extrae la serie de datos del marco de datos `df_can` correspondiente a Haití utilizando la lista «year».

In [38]:
year = list(map(str, range(1990, 2014)))

In [39]:
haiti=df_can.loc['Haiti', year]
haiti

,Haiti
1990,2379
1991,2829
1992,2399
1993,3655
1994,2100
1995,2014
1996,1955
1997,1645
1998,1295
1999,1439


### **Filtrado según un criterio**
Para filtrar el dataframe según una condición, basta con pasar dicha condición como un vector booleano.

Por ejemplo, filtremos el dataframe para mostrar los datos sobre los países asiáticos (AreaName = Asia).

In [41]:
# 1. Crear la serie booleana «condition»
condition = df_can['Continent'] == 'Asia'
print(condition)

Country
Afghanistan        True
Albania           False
Algeria           False
American Samoa    False
Andorra           False
                  ...  
Viet Nam           True
Western Sahara    False
Yemen              True
Zambia            False
Zimbabwe          False
Name: Continent, Length: 195, dtype: bool


In [42]:
# 2. Introducir esta condición en el dataFrame
df_can[condition]

,Continent,Region,DevName,1980,1981,1982,1983,1984,1985,1986,...,2006,2007,2008,2009,2010,2011,2012,2013,Total,Total1
Country,,,,,,,,,,,,,,,,,,,,,
Afghanistan,Asia,Southern Asia,Developing regions,16,39,39,47,71,340,496,...,3009,2652,2111,1746,1758,2203,2635,2004,58639,58639
Armenia,Asia,Western Asia,Developing regions,0,0,0,0,0,0,0,...,218,198,205,267,252,236,258,207,3310,3310
Azerbaijan,Asia,Western Asia,Developing regions,0,0,0,0,0,0,0,...,236,203,125,165,209,138,161,57,2649,2649
Bahrain,Asia,Western Asia,Developing regions,0,2,1,1,1,3,0,...,12,22,9,35,28,21,39,32,475,475
Bangladesh,Asia,Southern Asia,Developing regions,83,84,86,81,98,92,486,...,4014,2897,2939,2104,4721,2694,2640,3789,65568,65568
Bhutan,Asia,Southern Asia,Developing regions,0,0,0,0,1,0,0,...,10,7,36,865,1464,1879,1075,487,5876,5876
Brunei Darussalam,Asia,South-Eastern Asia,Developing regions,79,6,8,2,2,4,12,...,5,11,10,5,12,6,3,6,600,600
Cambodia,Asia,South-Eastern Asia,Developing regions,12,19,26,33,10,7,8,...,529,460,354,203,200,196,233,288,6538,6538
China,Asia,Eastern Asia,Developing regions,5123,6682,3308,1863,1527,1816,1960,...,33518,27642,30037,29622,30391,28502,33024,34129,659962,659962


In [43]:
# Podemos pasar varios criterios en la misma línea.
# Filtremos por AreaName = Asia y RegName = Asia Meridional

df_can[(df_can['Continent']=='Asia') & (df_can['Region']=='Southern Asia')]

# Nota: al utilizar los operadores «and» y «or», pandas requiere que utilicemos «&» y «|» en lugar de «and» y «or».
# No olvides encerrar las dos condiciones entre paréntesis.

,Continent,Region,DevName,1980,1981,1982,1983,1984,1985,1986,...,2006,2007,2008,2009,2010,2011,2012,2013,Total,Total1
Country,,,,,,,,,,,,,,,,,,,,,
Afghanistan,Asia,Southern Asia,Developing regions,16,39,39,47,71,340,496,...,3009,2652,2111,1746,1758,2203,2635,2004,58639,58639
Bangladesh,Asia,Southern Asia,Developing regions,83,84,86,81,98,92,486,...,4014,2897,2939,2104,4721,2694,2640,3789,65568,65568
Bhutan,Asia,Southern Asia,Developing regions,0,0,0,0,1,0,0,...,10,7,36,865,1464,1879,1075,487,5876,5876
India,Asia,Southern Asia,Developing regions,8880,8670,8147,7338,5704,4211,7150,...,33848,28742,28261,29456,34235,27509,30933,33087,691904,691904
Iran (Islamic Republic of),Asia,Southern Asia,Developing regions,1172,1429,1822,1592,1977,1648,1794,...,7480,6974,6475,6580,7477,7479,7534,11291,175923,175923
Maldives,Asia,Southern Asia,Developing regions,0,0,0,1,0,0,0,...,0,2,1,7,4,3,1,1,30,30
Nepal,Asia,Southern Asia,Developing regions,1,1,6,1,2,4,13,...,540,511,581,561,1392,1129,1185,1308,10222,10222
Pakistan,Asia,Southern Asia,Developing regions,978,972,1201,900,668,514,691,...,13127,10124,8994,7217,6811,7468,11227,12603,241600,241600
Sri Lanka,Asia,Southern Asia,Developing regions,185,371,290,197,1086,845,1838,...,4714,4123,4756,4547,4422,3309,3338,2394,148358,148358


*Recupera los datos en los que AreaName sea «África» y RegName sea «África Meridional».*
*Muestra el marco de datos y averigua cuántas entradas hay.*

In [45]:
df_can[(df_can['Continent']=='Africa')&(df_can['Region']=='Southern Africa')]

,Continent,Region,DevName,1980,1981,1982,1983,1984,1985,1986,...,2006,2007,2008,2009,2010,2011,2012,2013,Total,Total1
Country,,,,,,,,,,,,,,,,,,,,,
Botswana,Africa,Southern Africa,Developing regions,10,1,3,3,7,4,2,...,11,8,28,15,42,53,64,76,396,396
Lesotho,Africa,Southern Africa,Developing regions,1,1,1,2,7,5,3,...,0,4,1,8,7,1,0,6,107,107
Namibia,Africa,Southern Africa,Developing regions,0,5,5,3,2,1,1,...,19,13,26,14,16,23,24,83,320,320
South Africa,Africa,Southern Africa,Developing regions,1026,1118,781,379,271,310,718,...,1111,1200,1123,1188,1238,959,1243,1240,40568,40568
Swaziland,Africa,Southern Africa,Developing regions,4,1,1,0,10,7,1,...,7,5,6,10,3,13,17,39,188,188


### **Ordenación de valores de un DataFrame o una serie** <br>
La función `sort_values()` se utiliza para ordenar un DataFrame o una serie en función de una o varias columnas. <br>Especificar la columna o columnas por las que se desea ordenar y el orden (ascendente o descendente). A continuación se muestra la sintaxis para utilizarla:<br>

```df.sort_values(col_name, axis=0, ascending=True, inplace=False, ignore_index=False)```<br>

col_name: la(s) columna(s) por la(s) que se va a ordenar. <br>
axis: eje a lo largo del cual se va a ordenar. 0 para ordenar por filas (por defecto) y 1 para ordenar por columnas.<br>
ascending: para ordenar en orden ascendente (True, por defecto) o descendente (False).<br>
inplace: para realizar la operación de ordenación in situ (True) o devolver una copia ordenada (False, por defecto).<br>
ignore_index: para restablecer el índice tras la ordenación (True) o mantener los valores originales del índice (False, por defecto).<br>

Ordenemos el dataframe df_can por la columna «Total», en orden descendente, para averiguar cuáles son los cinco países que más han contribuido a la inmigración en Canadá.

In [50]:
df_can.sort_values(by='Total', axis=0, ascending=False, inplace=False, ignore_index=False).head(5)

,Continent,Region,DevName,1980,1981,1982,1983,1984,1985,1986,...,2006,2007,2008,2009,2010,2011,2012,2013,Total,Total1
Country,,,,,,,,,,,,,,,,,,,,,
India,Asia,Southern Asia,Developing regions,8880,8670,8147,7338,5704,4211,7150,...,33848,28742,28261,29456,34235,27509,30933,33087,691904,691904
China,Asia,Eastern Asia,Developing regions,5123,6682,3308,1863,1527,1816,1960,...,33518,27642,30037,29622,30391,28502,33024,34129,659962,659962
United Kingdom of Great Britain and Northern Ireland,Europe,Northern Europe,Developed regions,22045,24796,20620,10015,10170,9564,9470,...,7140,8216,8979,8876,8724,6204,6195,5827,551500,551500
Philippines,Asia,South-Eastern Asia,Developing regions,6051,5921,5249,4562,3801,3150,4166,...,18400,19837,24887,28573,38617,36765,34315,29544,511391,511391
Pakistan,Asia,Southern Asia,Developing regions,978,972,1201,900,668,514,691,...,13127,10124,8994,7217,6811,7468,11227,12603,241600,241600


*Averigua cuáles son los tres países que más contribuyeron a la inmigración en Canadá en el año 2010. <br> Muestra los nombres de los países junto con el número de inmigrantes de ese año.*

In [54]:
df_can.sort_values(by='2010', axis=0, ascending=False, inplace=False, ignore_index=False)
top3_2010 = df_can['2010'].head(3)
top3_2010


,2010
Country,
Afghanistan,1758
Albania,561
Algeria,4752
